storage/src/main/scala/coop/rchain/storage/regex/RegexPattern.scala

```diff
@@ -0,0 +1,364 @@
+      case (Some(x), Some(y)) => (0 <= x) && (x <= y)
+      case (Some(x), None)    => 0 <= x //min valid, max infinite
+      case (None, Some(_))    => false //min is infinite
+      case (None, None)       => true
+    },
+    "Invalid multiplier bounds"
+  )
+
+  /**
+    * We need some math operation on Option[Int], assuming that None == infinite
+    */
+  implicit class OptionIntMath(opt: Option[Int]) {
+    def +(that: Option[Int]): Option[Int] = (opt, that) match {
+      case (Some(_), None)    => None
+      case (None, Some(_))    => None
+      case (None, None)       => None
+      case (Some(x), Some(y)) => Some(x + y)
+    }
+
+    /**
+      * Subtract another bound from this one.
+      * Caution: this operation is not meaningful for all bounds.
+      */
+    def -(that: Option[Int]): Option[Int] = (opt, that) match {
+      case (Some(_), None) =>
+        throw new IllegalArgumentException("Can't substract infinity") //something - infinity => invalid operation
+      case (None, None)       => Some(0) //Infinity minus infinity is zero.
+      case (None, _)          => None //inf - anything => inf
+      case (Some(x), Some(y)) => Some(x - y)
+    }
+
+    def *(that: Option[Int]): Option[Int] = (opt, that) match {
+      case (None, _)          => None
+      case (_, None)          => None
+      case (Some(x), Some(y)) => Some(x * y)
+    }
+
+    def >=(that: Option[Int]): Boolean = (opt, that) match {
+      case (None, _)          => true
+      case (Some(_), None)    => false
+      case (Some(x), Some(y)) => x >= y
+    }
+
+    def <=(that: Option[Int]): Boolean = (opt, that) match {
+      case (_, None)          => true
+      case (None, Some(_))    => false
+      case (Some(x), Some(y)) => x <= y
+    }
+  }
+
+  def mandatory: Option[Int] = min
+  def optional: Option[Int] = max - min
+
+  override def toString: String = (min, max) match {
+    case (Some(_), Some(0)) =>
+      throw new ClassFormatError("Can't serialise a multiplier with max bound 0")
+    case (Some(0), Some(1))           => "?"
+    case (Some(1), Some(1))           => ""
+    case (Some(0), None)              => "*"
+    case (Some(1), None)              => "+"
+    case (Some(x), None)              => s"{$x,}"
+    case (Some(x), Some(y)) if x == y => s"{$x}"
+    case (Some(x), Some(y))           => s"{$x,$y}"
+  }
+
+  def isOne: Boolean = min.contains(1) && max.contains(1)
+
+  /**
+    * Multiplication is not well-defined for all pairs of multipliers because
+    * the resulting possibilities do not necessarily form a continuous range.
+    * For example:
+    * {0,x} * {0,y} = {0,x*y}
+    * {2} * {3} = {6}
+    * {2} * {1,2} = ERROR
+    * The proof isn't simple but suffice it to say that {p,p+q} * {r,r+s} is
+    * equal to {pr, (p+q)(r+s)} only if s=0 or qr+1 >= p. If not, then at least
+    * one gap appears in the range. The first inaccessible number is (p+q)r + 1.
+    */
+  def canMultiplyBy(that: Multiplier): Boolean =
```

Comment:
I think we should encode that in the `multiply` method's type with `Either` instead.

Reply:
Method is used for control flow, we get a result in any case.